src/V3SplitVar.cpp

```diff
@@ -0,0 +1,682 @@
+            AstConst* const msbp = constifyIfNot(dtypep->rangep()->msbp());
+            UASSERT_OBJ(lsbp, dtypep->rangep()->lsbp(), "must be constant");
+            UASSERT_OBJ(msbp, dtypep->rangep()->msbp(), "must be constant");
+            const vlsint32_t lsb = lsbp->toSInt(), msb = msbp->toSInt();
+            UASSERT_OBJ(lsb <= msb, dtypep->rangep(), "lsb must not greater than msb");
+            for (vlsint32_t i = 0; i <= msb - lsb; ++i) {
+                // const std::string name = varp->name() + "__BRA__" + AstNode::encodeNumber(i + lsb) + "__KET__";
+                // unpacked array is traced as var(idx).
+                const std::string name = varp->name() + AstNode::encodeName('(' + cvtToStr(i + lsb) + ')');
+                AstVar* const newp = new AstVar(varp->fileline(), varp->varType(), name, dtypep->subDTypep());
+                newp->trace(varp->isTrace());
+                insertp->addNextHere(newp);
+                if (newp->width() == 1) {  // no need to try splitting
+                    insertp = newp;
+                } else {
+                    newp->addNextHere(insertp = new AstPragma(varp->fileline(), AstPragmaType::SPLIT_VAR));
+                }
+                vars.push_back(newp);
+            }
+
+            // refer the split variable
+            for (size_t i = 0; i < it->second.size(); ++i) {
+                AstArraySel* selp = it->second[i];
+                AstVarRef* const vrefp = VN_CAST(selp->fromp(), VarRef);
+                AstConst* const indexp = VN_CAST(selp->bitp(), Const);
+                UASSERT_OBJ(vrefp && indexp, selp, "already checked");
+                const uint32_t idx = indexp->toUInt();
+                // V3Width:width() removes VAR_BASE attribute and make index 0-origin.
+                AstVarRef* const new_vref = new AstVarRef(selp->fileline(), vars.at(idx), vrefp->lvalue());
+                selp->replaceWith(new_vref);
+                selp->deleteTree();
+                VL_DANGLING(selp);
+            }
+            varp->unlinkFrBack()->deleteTree();
+            VL_DANGLING(varp);
+            ++m_numSplit;
+        }
+        m_refs.clear();  // done
+    }
+
+public:
+    SplitUnpackedVarVisitor(AstNetlist* nodep, bool firstRun)
+        : m_modp(NULL)
+        , m_numSplit(0)
+        , m_firstRun(firstRun) {
+        iterate(nodep);
+    }
+    ~SplitUnpackedVarVisitor() {
+        UASSERT(m_refs.empty(), "Don't forget to call split()");
+        if (m_firstRun) V3Stats::addStat("SplitVar, Split Unpacked Array", m_numSplit);
+    }
+    int numSplit() const { return m_numSplit; }
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // Check if the passed variable can be split.
+    // Even if this function returns true, the variable may not be split
+    // because the access to the variable cannot be determined statically.
+    static bool canSplit(const AstVar* nodep) {
+        if (AstNodeDType* dtypep = nodep->subDTypep()) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'dtypep' is assigned a value that is never used.](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4724658)